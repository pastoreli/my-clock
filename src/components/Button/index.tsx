import React, { ReactElement, useMemo } from 'react';
import styled, { useTheme } from 'styled-components/native';
import {
  TouchableOpacityProps,
  ActivityIndicator,
  Pressable,
  View,
} from 'react-native';
import Text from '../Text';

export type ButtonProps = TouchableOpacityProps & {
  loading?: boolean;
  prepend?: ReactElement;
  append?: ReactElement | string;
  block?: boolean;
  outlined?: boolean;
  color?: string;
  textColor?: string;
};

type ButtonPalletProps = {
  border: string;
  color: string;
  bg: string;
  opacity: number;
};

const Button: React.FC<ButtonProps> = ({
  children,
  loading = false,
  disabled,
  prepend,
  append,
  color,
  textColor,
  outlined,
  ...props
}) => {
  const theme = useTheme();

  const pallet = useMemo(() => {
    const buttonPallet: ButtonPalletProps = {
      bg: color || theme.primary,
      border: color || theme.primary,
      color: textColor || theme.lighthen,
      opacity: 1,
    };

    if (disabled) {
      buttonPallet.bg = theme.lighthen3;
      buttonPallet.border = theme.lighthen3;
      buttonPallet.opacity = 0.4;
      buttonPallet.color = theme.darken3;
      return buttonPallet;
    }

    if (outlined) {
      buttonPallet.opacity = 0.2;
      if (!textColor) {
        buttonPallet.color = color || theme.primary;
      }
      return buttonPallet;
    }

    return buttonPallet;
  }, [outlined, disabled, color, textColor, theme]);

  return (
    <Container
      activeOpacity={0.6}
      disabled={disabled || loading}
      testID="Button"
      android_ripple={{
        color: theme.lighthen3,
        borderless: false,
        radius: 22.5,
      }}
      borderColor={pallet.border}
      {...props}>
      <BgContainer bg={pallet.bg} opacity={pallet.opacity} />
      <Content>
        {loading ? (
          <ActivityIndicator
            testID="Button-Loading"
            color={textColor || theme.lighthen}
          />
        ) : (
          <>
            {Boolean(prepend) && prepend}
            <Text
              color={pallet.color}
              size="16"
              weight="semibold"
              testID="Button-Text">
              {children}
            </Text>
            {Boolean(append) && prepend}
          </>
        )}
      </Content>
    </Container>
  );
};

export default Button;

const Container = styled(Pressable)<{ block?: boolean; borderColor: string }>`
  position: relative;
  overflow: hidden;
  min-height: 40px;
  border-radius: 22.5px;
  background-color: transparent;
  align-items: center;
  justify-content: center;
  flex-direction: row;
  min-width: 150px;
  width: ${({ block }) => (block ? '100%' : 'auto')};
  align-self: center;
  margin-bottom: 10px;
  border: 1px solid ${({ borderColor }) => borderColor};
`;

const BgContainer = styled(View)<{ bg: string; opacity: number }>`
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: ${({ bg }) => bg};
  opacity: ${({ opacity }) => opacity};
`;

const Content = styled(View)`
  padding: 15px;
`;
