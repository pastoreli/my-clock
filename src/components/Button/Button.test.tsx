import { Text } from 'react-native';
import React from 'react';
import { render, fireEvent } from '../../utils/test/test-setup';
import Button from './index';

describe('Button Component', () => {
  test('normal', () => {
    const { getByTestId } = render(<Button>Press Me</Button>);
    expect(getByTestId('Button-Text').children).toContain('Press Me');
  });

  test('loading', () => {
    const { getByTestId } = render(<Button loading>Press Me</Button>);
    expect(getByTestId('Button-Loading')).toBeTruthy();
  });

  test('append & prepend', () => {
    const { getByTestId } = render(
      <Button
        block
        append={<Text testID="Append">Append</Text>}
        prepend={<Text testID="Prepend">Prepend</Text>}>
        Press Me
      </Button>,
    );
    expect(getByTestId('Button-Text').children).toContain('Press Me');
  });

  test('disabled', () => {
    let pressed = false;
    const handlePress = () => (pressed = true);
    const { getByTestId } = render(
      <Button onPress={handlePress} disabled={true}>
        Press Me
      </Button>,
    );
    fireEvent.press(getByTestId('Button'));
    expect(pressed).toBe(false);
    expect(getByTestId('Button-Text').children).toContain('Press Me');
  });

  test('outlined', () => {
    const { getByTestId } = render(<Button outlined>Press Me</Button>);
    fireEvent.press(getByTestId('Button'));
    expect(getByTestId('Button-Text').children).toContain('Press Me');
  });

  test('outlined with text color', () => {
    const { getByTestId } = render(
      <Button outlined textColor="#000">
        Press Me
      </Button>,
    );
    fireEvent.press(getByTestId('Button'));
    expect(getByTestId('Button-Text').children).toContain('Press Me');
  });

  test('set color', () => {
    const { toJSON } = render(<Button color="#f1f1f1">Press Me</Button>);
    const buttonBody = JSON.stringify(toJSON());
    expect(buttonBody.includes('#f1f1f1')).toBeTruthy();
  });
});
