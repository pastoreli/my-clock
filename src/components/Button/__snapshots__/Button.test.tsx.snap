// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component append & prepend 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  activeOpacity={0.6}
  block={true}
  borderColor="#3B83F3"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "#3B83F3",
        "borderRadius": 22.5,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
        "minHeight": 40,
        "minWidth": 150,
        "overflow": "hidden",
        "position": "relative",
        "width": "100%",
      },
    ]
  }
  testID="Button"
>
  <View
    bg="#3B83F3"
    opacity={1}
    style={
      [
        {
          "backgroundColor": "#3B83F3",
          "height": "100%",
          "opacity": 1,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        },
      ]
    }
  >
    <Text
      testID="Prepend"
    >
      Prepend
    </Text>
    <Text
      color="#FFFFFF"
      size="16"
      style={
        [
          {
            "color": "#FFFFFF",
            "fontFamily": "Montserrat-SemiBold",
            "fontSize": 16,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "textAlign": "auto",
            "textTransform": "none",
          },
        ]
      }
      testID="Button-Text"
      weight="semibold"
    >
      Press Me
    </Text>
    <Text
      testID="Prepend"
    >
      Prepend
    </Text>
  </View>
</View>
`;

exports[`Button Component disabled 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  activeOpacity={0.6}
  borderColor="#C6C9D8"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "#C6C9D8",
        "borderRadius": 22.5,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
        "minHeight": 40,
        "minWidth": 150,
        "overflow": "hidden",
        "position": "relative",
        "width": "auto",
      },
    ]
  }
  testID="Button"
>
  <View
    bg="#C6C9D8"
    opacity={0.4}
    style={
      [
        {
          "backgroundColor": "#C6C9D8",
          "height": "100%",
          "opacity": 0.4,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        },
      ]
    }
  >
    <Text
      color="#53555C"
      size="16"
      style={
        [
          {
            "color": "#53555C",
            "fontFamily": "Montserrat-SemiBold",
            "fontSize": 16,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "textAlign": "auto",
            "textTransform": "none",
          },
        ]
      }
      testID="Button-Text"
      weight="semibold"
    >
      Press Me
    </Text>
  </View>
</View>
`;

exports[`Button Component loading 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  activeOpacity={0.6}
  borderColor="#3B83F3"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "#3B83F3",
        "borderRadius": 22.5,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
        "minHeight": 40,
        "minWidth": 150,
        "overflow": "hidden",
        "position": "relative",
        "width": "auto",
      },
    ]
  }
  testID="Button"
>
  <View
    bg="#3B83F3"
    opacity={1}
    style={
      [
        {
          "backgroundColor": "#3B83F3",
          "height": "100%",
          "opacity": 1,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        },
      ]
    }
  >
    <ActivityIndicator
      color="#FFFFFF"
      testID="Button-Loading"
    />
  </View>
</View>
`;

exports[`Button Component normal 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  activeOpacity={0.6}
  borderColor="#3B83F3"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "#3B83F3",
        "borderRadius": 22.5,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
        "minHeight": 40,
        "minWidth": 150,
        "overflow": "hidden",
        "position": "relative",
        "width": "auto",
      },
    ]
  }
  testID="Button"
>
  <View
    bg="#3B83F3"
    opacity={1}
    style={
      [
        {
          "backgroundColor": "#3B83F3",
          "height": "100%",
          "opacity": 1,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        },
      ]
    }
  >
    <Text
      color="#FFFFFF"
      size="16"
      style={
        [
          {
            "color": "#FFFFFF",
            "fontFamily": "Montserrat-SemiBold",
            "fontSize": 16,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "textAlign": "auto",
            "textTransform": "none",
          },
        ]
      }
      testID="Button-Text"
      weight="semibold"
    >
      Press Me
    </Text>
  </View>
</View>
`;

exports[`Button Component outlined 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  activeOpacity={0.6}
  borderColor="#3B83F3"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "#3B83F3",
        "borderRadius": 22.5,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
        "minHeight": 40,
        "minWidth": 150,
        "overflow": "hidden",
        "position": "relative",
        "width": "auto",
      },
    ]
  }
  testID="Button"
>
  <View
    bg="#3B83F3"
    opacity={0.2}
    style={
      [
        {
          "backgroundColor": "#3B83F3",
          "height": "100%",
          "opacity": 0.2,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        },
      ]
    }
  >
    <Text
      color="#3B83F3"
      size="16"
      style={
        [
          {
            "color": "#3B83F3",
            "fontFamily": "Montserrat-SemiBold",
            "fontSize": 16,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "textAlign": "auto",
            "textTransform": "none",
          },
        ]
      }
      testID="Button-Text"
      weight="semibold"
    >
      Press Me
    </Text>
  </View>
</View>
`;

exports[`Button Component outlined with text color 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  activeOpacity={0.6}
  borderColor="#3B83F3"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "#3B83F3",
        "borderRadius": 22.5,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
        "minHeight": 40,
        "minWidth": 150,
        "overflow": "hidden",
        "position": "relative",
        "width": "auto",
      },
    ]
  }
  testID="Button"
>
  <View
    bg="#3B83F3"
    opacity={0.2}
    style={
      [
        {
          "backgroundColor": "#3B83F3",
          "height": "100%",
          "opacity": 0.2,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        },
      ]
    }
  >
    <Text
      color="#000"
      size="16"
      style={
        [
          {
            "color": "#000",
            "fontFamily": "Montserrat-SemiBold",
            "fontSize": 16,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "textAlign": "auto",
            "textTransform": "none",
          },
        ]
      }
      testID="Button-Text"
      weight="semibold"
    >
      Press Me
    </Text>
  </View>
</View>
`;

exports[`Button Component set color 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  activeOpacity={0.6}
  borderColor="#f1f1f1"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "#f1f1f1",
        "borderRadius": 22.5,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
        "minHeight": 40,
        "minWidth": 150,
        "overflow": "hidden",
        "position": "relative",
        "width": "auto",
      },
    ]
  }
  testID="Button"
>
  <View
    bg="#f1f1f1"
    opacity={1}
    style={
      [
        {
          "backgroundColor": "#f1f1f1",
          "height": "100%",
          "opacity": 1,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      [
        {
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        },
      ]
    }
  >
    <Text
      color="#FFFFFF"
      size="16"
      style={
        [
          {
            "color": "#FFFFFF",
            "fontFamily": "Montserrat-SemiBold",
            "fontSize": 16,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "textAlign": "auto",
            "textTransform": "none",
          },
        ]
      }
      testID="Button-Text"
      weight="semibold"
    >
      Press Me
    </Text>
  </View>
</View>
`;
